KeyboardMouseWin/Service/ElementLookupService.cs

```diff
@@ -0,0 +1,63 @@
+﻿using KeyboardMouseWin.Provider;
+using System.Collections.Concurrent;
+
+namespace KeyboardMouseWin.Service
+{
+    public class ElementLookupService : IElementLookupService
+    {
+        private readonly IUIElementProvider uIElementProvider;
+
+        public ElementLookupService(IUIElementProvider uIElementProvider)
+        {
+            this.uIElementProvider = uIElementProvider;
+        }
+        public async Task CaptionUiElementsAsync(IEnumerable<IUIElement> startingElements, Action<IEnumerable<IUIElement>>? elementsAddedAction = null, CancellationToken ct = default)
+        {
+            try
+            {
+                var updateTask = UpdateElementsAsync(new ConcurrentBag<IUIElement>(), startingElements, ct, elementsAddedAction);
+                var timeoutTask = Task.Delay(Timeout.Infinite, ct);
+
+                var completedTask = await Task.WhenAny(updateTask, timeoutTask);
+
+                if (completedTask == updateTask)
+                {
+                    // The updateTask completed first
+                    // Add your logic here
+                }
+                else if (completedTask == timeoutTask)
+                {
+                    // The timeoutTask completed first
+                    // Add your logic here
+                }
+            }
+            catch (TaskCanceledException ex)
+            {
+                // catch Cancellation token exception from Task.Run() and return
+            }
+
+        }
+        private async Task UpdateElementsAsync(ConcurrentBag<IUIElement> uIElements, IEnumerable<IUIElement> newElements, CancellationToken cancellationToken, Action<IEnumerable<IUIElement>>? elementsAddedAction = null)
```

Comment:
```csharp
        private async Task UpdateElementsAsync(ConcurrentBag<IUIElement> uIElements, IEnumerable<IUIElement> newElements, CancellationToken cancellationToken, Action<IEnumerable<IUIElement>>? elementsAddedAction = null)
        {
            if (cancellationToken.IsCancellationRequested) return;
            var queue = new Queue<IUIElement>(newElements);
            while (queue.Count > 0)
            {
                var element = queue.Dequeue();
                uIElements.Add(element);

                elementsAddedAction?.Invoke(uIElements);

                var subelements = uIElementProvider.GetSubElements(element);
                foreach (var subelement in subelements)
                {
                    queue.Enqueue(subelement);
                }

                if (cancellationToken.IsCancellationRequested) return;
                await Task.Yield();
            }
        }
```
This would be an alternative implementation